src/vdc_collection.rs

```diff
@@ -1,197 +1,193 @@
 use crate::common::*;
 use crate::storage_manager::*;
 
+use std::sync::Arc;
+
 use serde::{Deserialize, Serialize};
 use thiserror::Error;
 use tracing::{info, info_span};
 
 /// Internal prefix for credential keys.
 const KEY_PREFIX: &str = "Credential.";
 
-/// Supported credential formats.
-#[derive(PartialEq, Debug, Serialize, Deserialize)]
-pub enum CredentialFormat {
-    MsoMdoc,
-    JwtVcJson,
-    LdpVc,
-    Other(String), // For ease of expansion.
-}
-
-/// Supported credential types.
-#[derive(PartialEq, Debug, Serialize, Deserialize)]
-pub enum CredentialType {
-    Iso18013_5_1mDl,
-    VehicleTitle,
-    Other(String), // For ease of expansion.
-}
-
 /// An individual credential.
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug, uniffi::Object, Serialize, Deserialize)]
 pub struct Credential {
     id: Uuid,
-    format: CredentialFormat,
+    format: ClaimFormatDesignation,
     ctype: CredentialType,
     payload: Vec<u8>, // The actual credential.
 }
 
+#[uniffi::export]
 impl Credential {
-    /// Create a new credential.
-    fn new(
+    #[uniffi::constructor]
+    pub fn new(
         id: Uuid,
-        format: CredentialFormat,
+        format: ClaimFormatDesignation,
         ctype: CredentialType,
         payload: Vec<u8>,
-    ) -> Credential {
-        Credential {
+    ) -> Self {
+        Self {
             id,
             format,
             ctype,
             payload,
         }
     }
+
+    /// Get the ID of the credential.
+    pub fn id(&self) -> Uuid {
+        self.id.clone()
+    }
+
+    /// Get the format of the credential.
+    pub fn format(&self) -> ClaimFormatDesignation {
+        self.format.clone()
+    }
+
+    /// Get the type of the credential.
+    pub fn ctype(&self) -> CredentialType {
+        self.ctype.clone()
+    }
+
+    /// Get the payload of the credential.
+    pub fn payload(&self) -> Vec<u8> {
+        self.payload.clone()
+    }
+
+    /// Return the credential as a key suitable for storage.
+    pub fn as_storage_key(&self) -> Key {
+        Key(format!("{}{}", self.as_storage_key_prefix(), self.id))
+    }
+
+    /// Return the credential storage key prefix with credential type index as a string.
+    pub fn as_storage_key_prefix(&self) -> String {
+        format!("{KEY_PREFIX}{}.", self.ctype)
+    }
 }
 
 /// Verifiable Digital Credential Collection
 ///
 /// This is the main interface to credentials.
-pub struct VdcCollection {
-    storage: Box<dyn StorageManagerInterface>,
-}
+#[derive(Debug)]
+pub struct VdcCollection;
```

Comment:
What's the motivation for this change?

Reply:
I'm not Ryan, but I think the reasoning here was to avoid dealing with the memory management of passing around the Box. Since VdcCollection is used from the Wallet, which stores its own `StorageManagerInterface`, we can pass in a reference to the storage manager when we need it instead of storing it permanently in the VdcCollection. See the changes to `add`, `get`, etc below.

Reply:
It's not clear to me how this is a problem, it just moves the reference to the storage manager from the constructor to the methods. AFAIK there's no complexity with memory management - when the VdcCollection is constructed we have a reference to the storage manager, and when it goes out of scope / is GC'd, the reference is destroyed. Is there some nuance I'm missing?

Reply:
I'll let @Ryanmtate respond, but it also occurred to me that this might be a change related to moving away from using the callback interface from Uniffi, which required passing everything around in `Box`es, and instead being able to use regular borrows, but that storing a borrow in the struct itself would have caused an explosion of lifetimes. Again, I don't know for sure, though.